app/data_sources/azuredevops.py

```diff
@@ -0,0 +1,117 @@
+import datetime
+import logging
+from typing import Dict, List
+import requests
+import base64
+import urllib.parse
+
+from azure.devops.connection import Connection
+from msrest.authentication import BasicAuthentication
+
+from data_source_api.base_data_source import BaseDataSource, ConfigField, HTMLInputType
+from data_source_api.basic_document import DocumentType, BasicDocument
+from data_source_api.exception import InvalidDataSourceConfig
+from index_queue import IndexQueue
+from pydantic import BaseModel
+from parsers.html import html_to_text
+from data_source_api.utils import parse_with_workers
+
+logger = logging.getLogger(__name__)
+
+class DevOpsConfig(BaseModel):
+    organization_url: str
+    access_token: str
+    project_name: str
+    query_id: str
+
+class AzuredevopsDataSource(BaseDataSource):
+    @staticmethod
+    def get_config_fields() -> List[ConfigField]:
+        return [
+            ConfigField(label="AzureDevOps organization URL", placeholder="https://dev.azure.com/org", name="organization_url"),
+            ConfigField(label="Personal Access Token", name="access_token", type=HTMLInputType.PASSWORD),
+            ConfigField(label="Project Name", name="project_name"),
+            ConfigField(label="Query ID", name="query_id"),
+        ]
+    
+    @staticmethod
+    def validate_config(config: Dict) -> None:
+        try:
+            devops_config = DevOpsConfig(**config)
+            credentials = BasicAuthentication('', devops_config.access_token)
+            connection = Connection(base_url=devops_config.organization_url, creds=credentials)
+            core_client = connection.clients.get_core_client()
+            core_client.get_projects()
+        except Exception as e:
+            raise InvalidDataSourceConfig from e
+        
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        devops_config = DevOpsConfig(**self._config)
+        self._query_id = devops_config.query_id.strip()
+        self._access_token = devops_config.access_token.strip()
+        self._project_name = devops_config.project_name.strip()
+        self._organization_url = devops_config.organization_url.strip()
```

Comment:
could this be done as a `__post_init__` of the `DevOpsConfig` dataclass? should keep the code cleaner and make the validation aligned with the actual connection

Reply:
@Itaykal this has been adjusted.